mlflow/pyfunc/scoring_server/client.py

```diff
@@ -0,0 +1,76 @@
+import requests
+import time
+from mlflow.pyfunc import scoring_server
+import subprocess
+import os
+import sys
+import signal
+import json
+import atexit
+import warnings
+
+
+class ScoringServerClient:
+    def __init__(self, host, port):
+        self.url_prefix = f"http://{host}:{port}"
+
+    def ping(self):
+        ping_status = requests.get(url=self.url_prefix + "/ping")
+        if ping_status.status_code != 200:
+            raise Exception(f"ping failed (error code {ping_status.status_code})")
+
+    def wait_server_ready(self, timeout=30):
+        begin_time = time.time()
+        while True:
+            time.sleep(0.3)
+            try:
+                self.ping()
+                return
+            except Exception:
+                pass
+            if time.time() - begin_time > timeout:
+                break
+        raise RuntimeError("Wait scoring server ready timeout.")
+
+    def invoke(self, data, pandas_orient="records"):
+        """
+        Invoke inference on input data. The input data must be pandas dataframe or json instance.
+        """
+
+        import pandas as pd
+
+        content_type_list = [scoring_server.CONTENT_TYPE_JSON]
+        if isinstance(data, pd.DataFrame):
+            if pandas_orient == "records":
+                content_type_list.append(scoring_server.CONTENT_TYPE_FORMAT_RECORDS_ORIENTED)
+            elif pandas_orient == "split":
+                content_type_list.append(scoring_server.CONTENT_TYPE_FORMAT_SPLIT_ORIENTED)
+            else:
+                raise Exception(
+                    "Unexpected pandas_orient for Pandas dataframe input %s" % pandas_orient
+                )
+        post_data = json.dumps(scoring_server._get_jsonable_obj(data, pandas_orient=pandas_orient))
+
+        response = requests.post(
+            url=self.url_prefix + "/invocations",
+            data=post_data,
+            headers={"Content-Type": "; ".join(content_type_list)},
+        )
```

Comment:
Leaving this here again. It was marked `resolved` but doesn't seem resolved.

For general-purpose model scoring, `data` doesn't have to be a Pandas DataFrame. It could also be a list or a numpy array. Because this class isn't specific to Spark UDF, I think we should make this a generic interface that can accept any pyfunc-compatible input.

`scoring_server._get_jsonable_obj` already handles the conversion of pyfunc-compatible inputs to JSON. All we need to do is specify the `CONTENT_TYPE_JSON` content type. For an example, see https://github.com/mlflow/mlflow/blob/9781af9c0898827bf616a8f159168477a69036dd/mlflow/sagemaker/__init__.py#L2519-L2523.

Reply:
We probably don't need the `pandas_orient` argument here, since the server can handle both and the user likely doesn't care which format is used for the request.